[Codegen] Compilation error for DataObjects with "is" prefix for Booleans

Consider the following `DataObject`:
```kotlin
@DataObject(generateConverter = true)
class Account() {
  var isActive: Boolean? = null

  constructor(json: JsonObject) : this() {
    AccountConverter.fromJson(json, this)
  }
}
```

When the converter is generated it will incorrectly reference `active` instead of `isActive`:
```kotlin
fun accountOf(
  active: Boolean? = null): Account = io.vertx.example.Account().apply {

  if (active != null) {
    this.active = active
  }
}
```
...resulting in a compilation error:
```kotlin
> Task :compileKotlin FAILED
e: .../project/build/generated/source/kaptKotlin/main/io/vertx/example/Account.kt: (24, 10): Unresolved reference: active
e: .../project/build/generated/source/kaptKotlin/main/io/vertx/example/Account.kt: (36, 10): Unresolved reference: active

FAILURE: Build failed with an exception.
```
---

This seems to be due to the fact that `isActive: Boolean` translates into GET:`isActive()` and SET:`setActive()`, e.g. https://github.com/sockeqwe/fragmentargs/issues/46

--- a/vertx-lang-kotlin-gen/src/main/java/io/vertx/lang/kotlin/KotlinDataObjectGenerator.java
+++ b/vertx-lang-kotlin-gen/src/main/java/io/vertx/lang/kotlin/KotlinDataObjectGenerator.java
@@ -24,7 +24,6 @@ import io.vertx.codegen.annotations.ModuleGen;
 import io.vertx.codegen.doc.Doc;
 import io.vertx.codegen.doc.Token;
 import io.vertx.codegen.type.ClassKind;
-import io.vertx.codegen.type.ClassTypeInfo;
 import io.vertx.codegen.type.TypeInfo;
 import io.vertx.codegen.writer.CodeWriter;
 import io.vertx.lang.kotlin.helper.KotlinCodeGenHelper;
@@ -99,7 +98,7 @@ public class KotlinDataObjectGenerator extends KotlinGeneratorBase<DataObjectMod
       writer.println(" *");
       Token.toHtml(doc.getTokens(), " *", KotlinCodeGenHelper::renderLinkToHtml, "\n", writer);
       writer.println(" *");
-      model.getPropertyMap().values().stream().filter(filterProperties()).forEach(p -> {
+      model.getPropertyMap().values().stream().filter(filterReadOnlyProperties()).forEach(p -> {
         writer.print(" * @param " + p.getName() + " ");
         if (p.getDoc() != null) {
           String docInfo = Token.toHtml(p.getDoc().getTokens(), "", KotlinCodeGenHelper::renderLinkToHtml, "").replace("/*", "/<star>");
@@ -112,10 +111,6 @@ public class KotlinDataObjectGenerator extends KotlinGeneratorBase<DataObjectMod
       writer.println(" * NOTE: This function has been automatically generated from the [" + model.getType().getName() + " original] using Vert.x codegen.");
       writer.println(" */");
     }
-  }
-
-  private Predicate<PropertyInfo> filterProperties() {
-    return p -> p.getSetterMethod() != null || p.getAdderMethod() != null;
   }
 
   private String generateFun(DataObjectModel model, CodeWriter writer) {
@@ -131,7 +126,7 @@ public class KotlinDataObjectGenerator extends KotlinGeneratorBase<DataObjectMod
     String parameters = model.getPropertyMap()
       .values()
       .stream()
-      .filter(filterProperties())
+      .filter(filterReadOnlyProperties())
       .map(PropertyInfo::getName)
       .collect(Collectors.joining(", "));
     writer.print(parameters);
@@ -146,12 +141,11 @@ public class KotlinDataObjectGenerator extends KotlinGeneratorBase<DataObjectMod
   private void generateFunction(String functionName, DataObjectModel model, CodeWriter writer) {
     boolean isKotlin = model.getAnnotations().stream().anyMatch(ann -> ann.getName().equals("kotlin.Metadata"));
 
-    ClassTypeInfo type = model.getType();
     writer.println("fun " + functionName + "(");
     String paramsInfo = model.getPropertyMap()
       .values()
       .stream()
-      .filter(filterProperties())
+      .filter(filterReadOnlyProperties())
       .map(p -> {
         String type = mapKotlinType(p.getType());
         String name = isKotlin ? getKotlinName(p, type) : p.getName();
@@ -172,7 +166,7 @@ public class KotlinDataObjectGenerator extends KotlinGeneratorBase<DataObjectMod
     writer.indent();
     model.getPropertyMap().values()
       .stream()
-      .filter(filterProperties())
+      .filter(filterReadOnlyProperties())
       .forEach(p -> {
         String type = mapKotlinType(p.getType());
         String name = isKotlin ? getKotlinName(p, type) : p.getName();
@@ -204,6 +198,13 @@ public class KotlinDataObjectGenerator extends KotlinGeneratorBase<DataObjectMod
       });
     writer.unindent().println("}");
     writer.println();
+  }
+
+  /**
+   * Filter properties that are read-only (i.e. have no setter or adder method).
+   */
+  private Predicate<PropertyInfo> filterReadOnlyProperties() {
+    return p -> p.getSetterMethod() != null || p.getAdderMethod() != null;
   }
 
   /**

--- a/vertx-lang-kotlin-gen/src/main/java/io/vertx/lang/kotlin/KotlinDataObjectGenerator.java
+++ b/vertx-lang-kotlin-gen/src/main/java/io/vertx/lang/kotlin/KotlinDataObjectGenerator.java
@@ -152,7 +152,11 @@ public class KotlinDataObjectGenerator extends KotlinGeneratorBase<DataObjectMod
       .values()
       .stream()
       .filter(filterProperties())
-      .map(p -> "  " + p.getName() + ": " + applyPropertyKind(mapKotlinType(p.getType()), p.getKind()) + "? = null")
+      .map(p -> {
+        String type = mapKotlinType(p.getType());
+        String name = isKotlin ? getKotlinName(p, type) : p.getName();
+        return "  " + name + ": " + applyPropertyKind(type, p.getKind()) + "? = null";
+      })
       .collect(Collectors.joining(",\n"));
     writer.print(paramsInfo);
     writer.print("): ");
@@ -170,25 +174,26 @@ public class KotlinDataObjectGenerator extends KotlinGeneratorBase<DataObjectMod
       .stream()
       .filter(filterProperties())
       .forEach(p -> {
-        String propertyName = p.getName();
-        PropertyKind propertyKind = p.getKind();
-        writer.println("if (" + propertyName + " != null) {");
+        String type = mapKotlinType(p.getType());
+        String name = isKotlin ? getKotlinName(p, type) : p.getName();
+        PropertyKind kind = p.getKind();
+        writer.println("if (" + name + " != null) {");
         writer.indent();
         if (p.getSetterMethod() != null) {
-          writer.print("this." + (isKotlin ? (p.getName() + " = ") : (p.getSetterMethod() + "(")));
-          writer.print(propertyName);
-          if (propertyKind.isList()) {
+          writer.print("this." + (isKotlin ? (name + " = ") : (p.getSetterMethod() + "(")));
+          writer.print(name);
+          if (kind.isList()) {
             writer.print(".toList()");
-          } else if (propertyKind.isSet()) {
+          } else if (kind.isSet()) {
             writer.print(".toSet()");
           }
           writer.println((isKotlin ? "" : ")"));
         } else {
-          writer.print("for (item in " + propertyName);
+          writer.print("for (item in " + name);
           writer.println(") {");
           writer.indent();
           writer.print("this." + p.getAdderMethod());
-          if (propertyKind.isMap()) {
+          if (kind.isMap()) {
             writer.println("(item.key, item.value)");
           } else {
             writer.println("(item)");
@@ -199,6 +204,21 @@ public class KotlinDataObjectGenerator extends KotlinGeneratorBase<DataObjectMod
       });
     writer.unindent().println("}");
     writer.println();
+  }
+
+  /**
+   * Kotlin converts boolean fields to "isFoo" and "setFoo".
+   * Consequently, the field "isFoo" will incorrectly be mapped to "foo".
+   *
+   * @see <a href="https://github.com/vert-x3/vertx-lang-kotlin/issues/158">Releva</a>
+   */
+  private String getKotlinName(PropertyInfo p, String kotlinType) {
+    String name = p.getName();
+
+    if (kotlinType.equals("Boolean") && (p.getGetterMethod().startsWith("is") && !p.getSetterMethod().startsWith("is"))) {
+      name = p.getGetterMethod();
+    }
+    return name;
   }
 
   private String mapKotlinType(TypeInfo type) {

--- a/vertx-lang-kotlin-gen/src/test/kotlin/io/vertx/lang/kotlin/test/DataObjectBooleanTest.kt
+++ b/vertx-lang-kotlin-gen/src/test/kotlin/io/vertx/lang/kotlin/test/DataObjectBooleanTest.kt
@@ -1,0 +1,49 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Apache License v2.0 which accompanies this distribution.
+ *
+ * The Eclipse Public License is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * The Apache License v2.0 is available at
+ * http://www.opensource.org/licenses/apache2.0.php
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+package io.vertx.lang.kotlin.test
+
+import io.vertx.core.json.JsonObject
+import io.vertx.kotlin.lang.kotlin.test.dataobject.BooleanTestObject
+import io.vertx.kotlin.lang.kotlin.test.dataobject.booleanTestObjectOf
+import io.vertx.lang.kotlin.test.dataobject.BooleanTestObject
+import org.junit.Test
+
+class DataObjectBooleanTest {
+
+  @Test
+  fun `dataobject function has valid property for boolean`() {
+    var dataObject = BooleanTestObject(JsonObject())
+    dataObject = booleanTestObjectOf(
+      notPrefixed = true
+    )
+
+    dataObject = BooleanTestObject(
+      notPrefixed = true
+    )
+  }
+
+  @Test
+  fun `dataobject function has valid property for boolean with "is" prefix`() {
+    var dataObject = BooleanTestObject(JsonObject())
+    dataObject = booleanTestObjectOf(
+      isPrefixed = true
+    )
+
+    dataObject = BooleanTestObject(
+      isPrefixed = true
+    )
+  }
+}

--- a/vertx-lang-kotlin-gen/src/test/kotlin/io/vertx/lang/kotlin/test/dataobject/BooleanTestObject.kt
+++ b/vertx-lang-kotlin-gen/src/test/kotlin/io/vertx/lang/kotlin/test/dataobject/BooleanTestObject.kt
@@ -1,0 +1,15 @@
+package io.vertx.lang.kotlin.test.dataobject
+
+import io.vertx.codegen.annotations.DataObject
+import io.vertx.core.json.JsonObject
+
+@DataObject
+class BooleanTestObject {
+  var isPrefixed: Boolean? = null
+  var notPrefixed: Boolean? = null
+
+  constructor(json: JsonObject) {
+    isPrefixed = json.getBoolean("isPrefixed")
+    notPrefixed = json.getBoolean("notPrefixed")
+  }
+}

